// http-proxy-agent/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpProxyAgent = void 0;
const net = __importStar(require("net"));
const tls = __importStar(require("tls"));
const debug_1 = __importDefault(require("debug"));
const events_1 = require("events");
const agent_base_1 = require("agent-base");
const url_1 = require("url");
const debug = (0, debug_1.default)('http-proxy-agent');
/**
 * The `HttpProxyAgent` implements an HTTP Agent subclass that connects
 * to the specified "HTTP proxy server" in order to proxy HTTP requests.
 */
class HttpProxyAgent extends agent_base_1.Agent {
    constructor(proxy, opts) {
        super(opts);
        this.proxy = typeof proxy === 'string' ? new url_1.URL(proxy) : proxy;
        this.proxyHeaders = opts?.headers ?? {};
        debug('Creating new HttpProxyAgent instance: %o', this.proxy.href);
        // Trim off the brackets from IPv6 addresses
        const host = (this.proxy.hostname || this.proxy.host).replace(/^\[|\]$/g, '');
        const port = this.proxy.port
            ? parseInt(this.proxy.port, 10)
            : this.proxy.protocol === 'https:'
                ? 443
                : 80;
        this.connectOpts = {
            ...(opts ? omit(opts, 'headers') : null),
            host,
            port,
        };
    }
    addRequest(req, opts) {
        req._header = null;
        this.setRequestProps(req, opts);
        // @ts-expect-error `addRequest()` isn't defined in `@types/node`
        super.addRequest(req, opts);
    }
    setRequestProps(req, opts) {
        const { proxy } = this;
        const protocol = opts.secureEndpoint ? 'https:' : 'http:';
        const hostname = req.getHeader('host') || 'localhost';
        const base = `${protocol}//${hostname}`;
        const url = new url_1.URL(req.path, base);
        if (opts.port !== 80) {
            url.port = String(opts.port);
        }
        // Change the `http.ClientRequest` instance's "path" field
        // to the absolute path of the URL that will be requested.
        req.path = String(url);
        // Inject the `Proxy-Authorization` header if necessary.
        const headers = typeof this.proxyHeaders === 'function'
            ? this.proxyHeaders()
            : { ...this.proxyHeaders };
        if (proxy.username || proxy.password) {
            const auth = `${decodeURIComponent(proxy.username)}:${decodeURIComponent(proxy.password)}`;
            headers['Proxy-Authorization'] = `Basic ${Buffer.from(auth).toString('base64')}`;
        }
        if (!headers['Proxy-Connection']) {
            headers['Proxy-Connection'] = this.keepAlive
                ? 'Keep-Alive'
                : 'close';
        }
        for (const name of Object.keys(headers)) {
            const value = headers[name];
            if (value) {
                req.setHeader(name, value);
            }
        }
    }
    async connect(req, opts) {
   )"/>
		</xsl:variable>
		<xsl:variable name="cYear">
			<xsl:value-of select="count(b:Year)"/>
		</xsl:variable>
		<xsl:variable name="cYearAccessed">
			<xsl:value-of select="count(b:YearAccessed)"/>
		</xsl:variable>
		<xsl:variable name="cURL">
			<xsl:value-of select="count(b:URL)"/>
		</xsl:variable>

		<xsl:variable name="SourceType">
			<xsl:value-of select="b:SourceType"/>
		</xsl:variable>

		<xsl:choose>
			<xsl:when test="$SourceType = 'Book'">
				<xsl:element name="tr">
					<xsl:element name="td">
						<xsl:attribute name="valign">
							<xsl:value-of select="'top'"/>
						</xsl:attribute>
						<xsl:attribute name="width">
							<xsl:value-of select="'1%'"/>
						</xsl:attribute>
						<xsl:element name="p">
							<xsl:call-template name = "BibAddParagraphAttributes"/>
							<xsl:call-template name="BibRefOrder"/>
						</xsl:element>
					</xsl:element>
					<xsl:element name="td">
						<xsl:attribute name="style">
							<xsl:value-of select="'text-align:left'"/>
						</xsl:attribute>
						<xsl:attribute name="valign">
							<xsl:value-of select="'top'"/>
						</xsl:attribute>
						<xsl:element name="p">
							<xsl:call-template name = "BibAddParagraphAttributes"/>
							<xsl:call-template name ="BibDisplayAuthorBook">
								<xsl:with-param name ="DisplayEditorIfAuthorUnavailale" select="'true'" />
							</xsl:call-template>
							<xsl:call-template name = "BibDisplayTitleBook"/>
							<xsl:call-template name = "BibDisplayEdition"/>
							<xsl:call-template name="BibDisplayVolumeBook"/>
							<xsl:call-template name="BibDisplayEditorBook"/>

							<xsl:call-template name = "BibDisplayCity"/>
							<xsl:call-template name = "BibDisplayStateProvinceBook"/>

							<xsl:call-template name = "BibDisplayPublisherBC"/>
							<xsl:call-template name = "BibDisplayYearBC"/>
							<xsl:call-template name ="BibDisplayPages"/>
						</xsl:element>
					</xsl:element>
				<
